### PR TITLE
Fix hessian2 serialized short, byte is converted to int bug

### PR DESCRIPTION
**Pull request by zonghaishang**
Fix hessian2 serialized short, byte is converted to int bug

for example:
 if we test deserialize mode 'Hessian2StringShortType'
`
public class Hessian2StringShortType implements Serializable {

    Map<String, Short> stringShortMap;

    Map<String, Byte> stringByteMap;

    Map<String, PersonType> stringPersonTypeMap;

    public Hessian2StringShortType(){

    }
}

public class PersonType implements Serializable {

    String name;
    int age;
    double money;
    short p1;
    byte  p2;
    List<Short> p3;

    public PersonType(String name, int age, double money, short p1, byte p2, List<Short> p3) {
        this.name = name;
        this.age = age;
        this.money = money;
        this.p1 = p1;
        this.p2 = p2;
        this.p3 = p3;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        PersonType type = (PersonType) o;
        return age == type.age &&
            Double.compare(type.money, money) == 0 &&
            p1 == type.p1 &&
            p2 == type.p2 &&
            Objects.equals(name, type.name) &&
            Objects.equals(p3, type.p3);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, age, money, p1, p2, p3);
    }
}
`
`
    @Test
    public void serialize_string_short_map_then_deserialize() throws Exception {

        Hessian2StringShortType stringShort = new Hessian2StringShortType();
        Map<String, Short> stringShortMap = new HashMap<String, Short>();
        stringShortMap.put("first", (short)0);
        stringShortMap.put("last", (short)60);
        stringShort.stringShortMap = stringShortMap;

        ByteArrayOutputStream bout = new ByteArrayOutputStream();
        Hessian2Output out = new Hessian2Output(bout);

        out.writeObject(stringShort);
        out.flush();

        ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
        Hessian2Input input = new Hessian2Input(bin);

        Hessian2StringShortType deserialize = (Hessian2StringShortType) input.readObject();
        assertTrue(deserialize.stringShortMap != null);
        assertTrue(deserialize.stringShortMap.size() == 2);
        // however, the deserialize result is Integer !!!
        assertTrue(deserialize.stringShortMap.get("last") instanceof Short);
        assertEquals(Short.valueOf((short)0), deserialize.stringShortMap.get("first"));
        assertEquals(Short.valueOf((short)60), deserialize.stringShortMap.get("last"));
    }
`
When we test serialize and deserialize 'Hessian2StringShortType', because it contains short、byte, 
the property stringShortMap will be deserialized Map<String, Integer>, not Map<String, Short>!!!

More detail unit test can see: `com.alibaba.com.caucho.hessian.io.Hessian2StringShortTest`

The problem has been fixed in the current branch, please verify and merge, Thank you.
Here is my email: shangzonghai@youzan.com or jason.shang@hotmail.com

### CODE REVIEW

**Comment by chickenlj** [comment]
@lovepoem Do you have time to review this PR? I've seen you fixed bugs in hessian before, you must be familiar with hessian code.

**Comment by lovepoem** [comment]
ok I will review this PR

**Comment by zonghaishang** [comment]
@lovepoem Do you have any question  about this pr ?

File: hessian-lite/src/test/java/com/alibaba/com/caucho/hessian/io/PersonType.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package com.alibaba.com.caucho.hessian.io;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Objects;
```
[the remark this answers is not in the document]
Comment (zonghaishang) [reply hessian-lite/src/test/java/com/alibaba/com/caucho/hessian/io/PersonType.java]: I‘m using Objects.hash is java7+ feature, must adapt jdk1.6 ?

**Comment by zonghaishang** [comment]
@chickenlj @lovepoem please help review code, thanks.

File: hessian-lite/src/main/java/com/alibaba/com/caucho/hessian/io/AbstractHessianInput.java [hunk 2]
```diff
@@ -334,12 +335,34 @@ abstract public InputStream readInputStream()
...
             throws IOException;
 
+    /**
+     * Reads an arbitrary object from the input stream.
+     *
+     * @param expectedClass the expected class if the protocol doesn't supply it.
+     * @param expectedTypes the runtime type hints, eg: expectedClass equals Map, expectedTypes can
+     * equals String.class, Short.class
+     */
+    public Object readObject(Class expectedClass, Class<?>... expectedTypes)
```
Comment (takeseem) [inline hessian-lite/src/main/java/com/alibaba/com/caucho/hessian/io/AbstractHessianInput.java]: the add method will work in Hessian2Input, but in HessianInput not unsupported. 
I think the HessianInput need to change to support the new methods.

**Comment by takeseem** [comment]
HessianInput.readObject(Class, Class[]) nosupported, effect #1608

**Pull request merged**